### PR TITLE
MULE-18906: Watermark in polling sources is updated when source is stopping.

### PR DESCRIPTION
**Pull request by SebaElizalde**

### CODE REVIEW

File: modules/extensions-support/src/main/java/org/mule/runtime/module/extension/internal/runtime/source/poll/PollingSourceWrapper.java [hunk 1]
```diff
@@ -603,14 +607,19 @@ private void setCurrentWatermarkAsMinimumRejectWatermark(Serializable minimumRej
...
     try {
       osClearingLock.lock();
       List<String> strings = idsOnUpdatedWatermark.allKeys();
-      recentlyProcessedIds.clear();
+      if (clearRecentlyProcessed) {
+        recentlyProcessedIds.clear();
+      }
       strings.forEach(key -> {
         try {
+          if (!clearRecentlyProcessed && recentlyProcessedIds.contains(key)) {
```
Comment (SebaElizalde) [inline modules/extensions-support/src/main/java/org/mule/runtime/module/extension/internal/runtime/source/poll/PollingSourceWrapper.java:620]: Comment from @Bardammu :
"We can try to retrieve all the values of recentlyProcessedIds and then call contains for each key. In this way, we avoid making a request for each key to OSv2.
Not in this PR, it's something we should think about."

File: tests/test-extensions/petstore-extension/src/main/java/org/mule/test/petstore/extension/WatermarkingPetAdoptionSource.java [hunk 2]
```diff
@@ -22,22 +22,22 @@
...
-  @Parameter
-  @org.mule.runtime.extension.api.annotation.param.Optional(defaultValue = "false")
-  protected boolean useWatermark;
+  public static CountDownLatch beginLatch = new CountDownLatch(1);
 
   @Parameter
-  @org.mule.runtime.extension.api.annotation.param.Optional(defaultValue = "false")
-  protected boolean idempotent;
+  @org.mule.runtime.extension.api.annotation.param.Optional(defaultValue = "0")
+  protected Integer awaitOnItem;
```
[the remark this answers is not in the document]
Comment (SebaElizalde) [reply tests/test-extensions/petstore-extension/src/main/java/org/mule/test/petstore/extension/WatermarkingPetAdoptionSource.java:40]: 0 means never wait, I can set it to -1 to be more clear. Or start the count from 0 (that would be the first item) and make -1 default.

[the remark this answers is not in the document]
Comment (ndinu) [reply tests/test-extensions/petstore-extension/src/main/java/org/mule/test/petstore/extension/WatermarkingPetAdoptionSource.java:40]: 👍 I think it is ok the way it is.

File: modules/extensions-spring-support/src/test/java/org/mule/test/module/extension/source/PollingSourceRestartingTestCase.java [hunk 3]
```diff
@@ -0,0 +1,166 @@
...
+    assertWatermarkingForStopStartScenario(asList("Anibal", "Barbara", "ANIBAL", "BARBARA", "Colonel Meow"),
+                                           "processedItemsWithNewWatermarkAreReprocessedWhenSourceIsRestartedMidPollWithLimit");
+    assertLimitIsApplied(2);
+  }
+
+  private void assertWatermarkingForStopStartScenario(List<String> expectedPets, String flowName) throws Exception {
+    startFlow(flowName);
+    beginLatch.await();
+    stopFlow(flowName);
+    check(5000, 1000, () -> getFlowConstruct(flowName).getLifecycleState().isStopped());
```
[the remark this answers is not in the document]
Comment (ndinu) [reply modules/extensions-spring-support/src/test/java/org/mule/test/module/extension/source/PollingSourceRestartingTestCase.java:127]: 😢

**Review by ndinu — APPROVED** [review]

**Review by ndinu — APPROVED** [review]

**Pull request merged**